presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyAggregation.java

```diff
@@ -0,0 +1,201 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.differentialentropy;
+
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.function.AggregationFunction;
+import com.facebook.presto.spi.function.AggregationState;
+import com.facebook.presto.spi.function.CombineFunction;
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.InputFunction;
+import com.facebook.presto.spi.function.OutputFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.type.StandardTypes;
+import io.airlift.slice.Slice;
+
+import java.security.InvalidParameterException;
+
+import static com.facebook.presto.operator.aggregation.differentialentropy.Constants.FIXED_HISTOGRAM_JACKNIFE_METHOD_NAME;
+import static com.facebook.presto.operator.aggregation.differentialentropy.Constants.FIXED_HISTOGRAM_MLE_METHOD_NAME;
+import static com.facebook.presto.operator.aggregation.differentialentropy.Constants.RESERVOIR_VASICEK_METHOD_NAME;
+import static com.facebook.presto.operator.aggregation.differentialentropy.Constants.WEIGHTED_RESERVOIR_VASICKE_METHOD_NAME;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.google.common.base.Verify.verify;
+import static java.lang.String.format;
+
+@AggregationFunction("differential_entropy")
+@Description("Computes differential entropy based on random-variable samples")
+public final class DifferentialEntropyAggregation
+{
+    private DifferentialEntropyAggregation() {}
+
+    @InputFunction
+    public static void input(
+            @AggregationState DifferentialEntropyState state,
+            @SqlType(StandardTypes.BIGINT) long size,
+            @SqlType(StandardTypes.DOUBLE) double sample,
+            @SqlType(StandardTypes.DOUBLE) double weight,
+            @SqlType(StandardTypes.VARCHAR) Slice method,
+            @SqlType(StandardTypes.DOUBLE) double min,
+            @SqlType(StandardTypes.DOUBLE) double max)
+    {
+        String requestedMethod = method == null ? null : method.toStringUtf8();
+        if (state.getStrategy() == null) {
+            DifferentialEntropyStateStrategy strategy;
+            if (requestedMethod.equalsIgnoreCase(FIXED_HISTOGRAM_MLE_METHOD_NAME)) {
+                strategy = new FixedHistogramMLEStateStrategy(size, min, max);
+            }
+            else if (requestedMethod.equalsIgnoreCase(FIXED_HISTOGRAM_JACKNIFE_METHOD_NAME)) {
+                strategy = new FixedHistogramJacknifeStateStrategy(size, min, max);
+            }
+            else {
+                throw new InvalidParameterException(format("unknown method %s or method/argument mismatch", method));
```

Comment:
This should be a PrestoException. Please, add a test.

`throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Invalid input to differential_entropy: unknown method: %s", method));`

Reply:
Added tests.